lib/ptracer.h

```diff
@@ -0,0 +1,485 @@
+    *mwhite = *mblack = 0;
+
+    for(int i = 0; i < cont->count; i++) {
+        *mwhite += (cont->stack[i].type == PALLENE_TRACER_FRAME_TYPE_C);
+        *mblack += (cont->stack[i].type == PALLENE_TRACER_FRAME_TYPE_LUA);
+    }
+}
+
+/* Responsible for printing and controlling some of the traceback fn parameters. */
+static void _pallene_tracer_dbg_print(const char *buf, bool *ellipsis, int *pframes, int nframes) {
+    /* We have printed the frame, even tho it might not be visible ;). */
+    (*pframes)++;
+
+    /* Should we print? Are we at any point in top or bottom printing threshold? */
+    bool should_print = (*pframes <= PALLENE_TRACEBACK_TOP_THRESHOLD)
+        || ((nframes - *pframes) <= PALLENE_TRACEBACK_BOTTOM_THRESHOLD);
+
+    if(luai_likely(should_print == true))
+        fprintf(stderr, buf);
+    else if(*ellipsis) {
+        fprintf(stderr, "\n    ... (Skipped %d frames) ...\n\n",
+            nframes - (PALLENE_TRACEBACK_TOP_THRESHOLD
+            + PALLENE_TRACEBACK_BOTTOM_THRESHOLD));
+
+        *ellipsis = false;
+    }
+}
+
+/* Frees the heap-allocated resources. */
+/* This function will be used as the `__gc` metamethod. */
+static int _pallene_tracer_free_resources(lua_State *L) {
+    pt_cont_t *cont = (pt_cont_t *) lua_touserdata(L, 1);
+
+    free(cont->stack);
+
+    return 0;
+}
+
+/* ---------------- PRIVATE END ---------------- */
+
+/* ---------------- DEFINITIONS ---------------- */
+
+/* Initializes the Pallene Tracer. The initialization refers to creating the stack
+   if not created, preparing the traceback fn and finalizers. */
+/* NOTE: Pushes the finalizer object to the stack. The object has to be closed
+   everytime you are in a Lua C function using `lua_toclose(L, idx)` function. */
+pt_cont_t *pallene_tracer_init(lua_State *L) {
+    pt_cont_t *cont = NULL;
+
+    /* Try getting the userdata. */
+    lua_getfield(L, LUA_REGISTRYINDEX, PALLENE_TRACER_CONTAINER_ENTRY);
+
+    /* If we don't find any userdata, initialize resources. */
+    if(luai_unlikely(lua_isnil(L, -1) == 1)) {
+        cont = (pt_cont_t *) lua_newuserdata(L, sizeof(pt_cont_t));
+        cont->stack = malloc(PALLENE_TRACER_MAX_CALLSTACK * sizeof(pt_frame_t));
+        cont->count = 0;
+
+        /* Prepare the `__gc` finalizer to free the stack. */
+        lua_newtable(L);
+        lua_pushcfunction(L, _pallene_tracer_free_resources);
+        lua_setfield(L, -2, "__gc");
+        lua_setmetatable(L, -2);
+
+        /* This is our finalizer which will reside in the value stack. */
+        lua_newtable(L);
+        lua_newtable(L);
+        lua_pushvalue(L, -3);
+
+        /* Our finalizer fn. */
+        lua_pushcclosure(L, pallene_tracer_finalizer, 1);
+        lua_setfield(L, -2, "__close");
+        lua_setmetatable(L, -2);
+
+        /* Set finalizer object to registry. */
+        lua_setfield(L, LUA_REGISTRYINDEX, PALLENE_TRACER_FINALIZER_ENTRY);
```

Comment:
Instead of using a string and hoping that it does not show up elsewhere, we can we use a stringified static variable.
From the docs:

> because all C libraries share the same registry, you must choose with care what values you use as keys, to avoid collisions. A bulletproof method is to use as key the address of a static variable in your code: The C link editor ensures that this key is unique among all libraries.

(Source: https://www.lua.org/pil/27.3.1.html)

Reply:
@hugomg and I have discussed about it. This would work if and only if the static variable we are using have a single instance throughout the Lua process (that is either only one module have this static variable or multiple module sharing this static variable through another shared library which is going to create a single instance of hte static variable in in the process).

But in Pallene tracer, you have the choice of statically or dynamically link to the library. If we choose to dynamically link the library, we are good to go. But the problem arises if we try to statically link the library. Statically linking the library (or including all the implementation functions and variables per se), the static variable we would use as a key would share multiple instances, thus having different addresses. Which basically won't work for our case.

As for why I decided to go with static linking alongside with dynamic linking is because `pallene-tracer` is a small library and not everyone would want to link it dynamically (people may have their reasons, personally I find it less of a hassle linking statically). On top of that, we don't need to deal with `.a` files for static linking, so it's even easier.